pkg/edgeview/src/copyfile.go

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type copyFile struct {
+	TokenHash []byte   `json:"tokenHash"`
+	Name      string   `json:"name"`
+	Size      int64    `json:"size"`
+	Sha256    string   `json:"sha256"`
+	ModTsec   int64    `json:"modtsec"`
+}
+
+type fileCopyStatus struct {
+	gotFileInfo   bool
+	filename      string
+	fileSize      int64
+	fileHash      string
+	currSize      int64
+	modTime       time.Time
+	buf           []byte
+	f             *os.File
+}
+
+func runCopy(opt string) {
+	path := strings.SplitN(opt, "cp/", 2)
+	if len(path) != 2 {
+		fmt.Printf("cp needs a cp/ and path input\n")
+		return
+	}
+	file := path[1]
+	info, err := os.Stat(file)
+	if err != nil {
+		fmt.Printf("os stat error %v\n", err)
+		return
+	}
+	//fmt.Printf("file info %+v\n", info)
+	cfile := copyFile{
+		TokenHash: tokenHash16,
+		Name:      info.Name(),
+		Size:      info.Size(),
+		ModTsec:   info.ModTime().Unix(),
+		Sha256:    fmt.Sprintf("%x", getFileSha256(file)),
+	}
+	jbytes, err := json.Marshal(cfile)
+	if err != nil {
+		fmt.Printf("json marshal error %v\n", err)
+		return
+	}
+	_ = websocketConn.WriteMessage(websocket.BinaryMessage, jbytes)
+
+	// server side set
+	isSvrCopy = true
+	copyMsgChn = make(chan []byte)
+	//websocketConn.SetReadDeadline(time.Now().Add(60 * time.Second))
+	ahead := make(chan struct{})
+	done := make(chan struct{})
+	t := time.NewTimer(30 * time.Second)
+	readerRunning := true
+
+	go func() {
+		for {
+			select {
+			case message := <-copyMsgChn:
+				if !strings.Contains(string(message), startCopyMessage) {
```

Comment:
Shouldn't these global variables copyMsgChn, startCopyMessage be defined in the current file instead of edge-view.go? It is a bit hard to read with usage and definitions in different files. Made me think that the file copy starts without the receiver acknowledging receipt of file information first.

Reply:
I was debating on to group them in one place or spreading them across. will think about this more.